quickcomm/models.py

```diff
@@ -18,6 +24,145 @@
+    # hash them, but for now, we will leave them as is.
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+
+    username = models.CharField(max_length=100, help_text="The username to use to authenticate to the host.", verbose_name="Username")
+    password = models.CharField(max_length=100, help_text="The password to use to authenticate to the host.", verbose_name="Password")
+    nickname = models.CharField(max_length=100, help_text="The nickname of the host authenticator. This is only used for display purposes.", verbose_name="Nickname", null=True, blank=True)
+
+
+    @property
+    def nickname_or_username(self):
+        if self.nickname:
+            return self.nickname
+        return self.username
+
+    @property
+    def base64string(self):
+        return base64.b64encode(f"{self.username}:{self.password}".encode('ascii')).decode('ascii')
+
+    def check_password(self, password):
+        return self.password == password
+
+    @property
+    def is_authenticated(self):
+        return True
+
+    def __str__(self):
+        return f"{self.username}"
+
+class Host(models.Model):
+    """A host is a remote server that hosts authors."""
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+
+    url = models.URLField(validators=[URLValidator], help_text="The URL of the host API. This must be in proper form (e.g. https://example.com/).", verbose_name="Host URL")
+
+    username_password_base64 = models.CharField(max_length=100, help_text="The username and password to use to authenticate to the host. This is a base64 encoded string of the form 'username:password'.", verbose_name="Username and Password", null=True, blank=True)
+
+    last_successful_ping = models.DateTimeField(null=True, blank=True)
+    last_ping = models.DateTimeField(null=True, blank=True)
+    last_ping_result = models.BooleanField(null=True, blank=True)
+
+    nickname = models.CharField(max_length=100, help_text="The nickname of the host. This is only used for display purposes.", verbose_name="Nickname", null=True, blank=True)
+
+    @property
+    def nickname_or_url(self):
+        if self.nickname:
+            return self.nickname
+        return self.url
+
+    def sync(self):
+        """Syncs the host with the remote API."""
+        self.sync_authors()
+
+    def sync_authors(self, admin=None, request=None):
+        """Syncs the authors with the remote API."""
+
+        try:
+            ext_authors = get_authors(self.url)
+            for ext_author in ext_authors:
+                self.deserialize_author(ext_author).save()
+        except Exception as e:
+            if admin is not None and request is not None:
+                admin.message_user(request, "Error syncing authors: {}".format(e))
+
+        # TODO delete authors that are no longer on the remote host
+        # TODO don't delete posts that were private by accident
```

Comment:
I think we should worry about this kind of logic later (i.e. most edge cases related to deleting/updates on external servers), but maybe create some issues for it so we know to come back to it next week